rfc/005-soft-chain-upgrades.md

```diff
@@ -0,0 +1,473 @@
+There may still remain the need for hard upgrades but this will be focused
+[elsewhere](https://github.com/tendermint/tendermint/issues/5595)). Instead,
+this RFC will describe what supporting soft upgrades would entail and explore
+two different methods: **Multi Data Structure Support** or
+**Chain Migrations**.
+
+## Proposal
+
+Tendermint supports a class of upgrades deemed as soft upgrades.
+
+A soft upgrade, in the simplest terms, is a change to Tendermint's binary that
+would not impede with the use of prior data / existing chains.
+
+The implementation to support this would ideally need to be done prior to the
+1.0 release.
+
+### Stakeholders
+
+Let's first set out to define the different needs that the various stakeholders
+of the Tendermint software have in the context of upgrading:
+
+- Application Developers: want to be able to get as much of the benefits of
+  new features (like state sync and light client) with minimal work or loss to
+  their current infrastructure. Ideally, Tendermint is also somewhat
+  accommodating to application upgrades.
+
+- Node Operators: Primarily they want safety and reliability. This means minimal
+  down-time and intuitive / easy UX when it comes to upgrading.
+
+- Wallets, Block Explorers and other clients: data retrievability which means
+  ensuring that helpful/informative data is always available and that minimal
+  infrastructure is needed to support serving data across the entire lifespan of
+  the chain.
+
+### Protocol Versioning
+
+Tendermint currently has three protocol versions: Block, P2P and App.
+
+The block protocol version would initially be the sole focus for soft upgrades.
+This consists of all block related data structures i.e. `Header`,
+`Vote`, `Validator`, `Commit` (a full list of the exact data structures can be
+found in [data structures](https://github.com/tendermint/spec/blob/master/spec/core/data_structures.md#data-structures).
+Data structures are dictated by the spec thus requiring a Tendermint software
+version to indicate the block version/s it supports. When a new spec release
+makes changes to any of the aforementioned data structures, this will result in
+incrementing the block protocol version.
+
+The app version can already be changed via consensus and is left to the
+discretion of the application to handle. The P2P version would also be left as
+it is, with major revisions requiring a hard upgrade. However, this may be
+revisited in the future.
+
+### Transitions
+
+Networks would be expected to announce upgrades in advance allowing a window for
+node operators to upgrade to a software version of Tendermint that would
+comply with the announced block protocol upgrade. This can be done
+asynchronously with node operators temporarily halting their node, changing to
+the new binary and starting up again. Tendermint could also offer a tool to
+easily swap between binaries thus reducing the amount of down-time that nodes
```

Comment:
I guess this is basically what Regen has built in the https://github.com/cosmos/cosmos-sdk/tree/master/cosmovisor ?

Reply:
Yup I was thinking of the cosmosvisor when I wrote this (although I'm not too familiar with it)